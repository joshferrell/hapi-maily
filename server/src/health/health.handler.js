import boom from 'boom';

export const handleCheckServer = (request, reply) => reply({ status: 'OK' });

/**
 * check dependencies for services
 * @param  {[promise functions]} promiseFunctions functions that when execute return promises
 * @return {reply}               replies result of promise functions
 */
export const makeHandleCheckDependency = promiseFunctions => (request, reply) =>
    Promise
        .all(promiseFunctions.map(x => x()))
        .then((dependencies) => {
            dependencies
                .filter(({ up }) => !up)
                .forEach(({ name }) => request.log('error', `dependency ${name} is down`));

            return reply(dependencies);
        })
        .catch((err) => {
            request.log('fatal', { err, msg: 'unable to test dependency status' });
            return reply(boom.badImplementation());
        });
