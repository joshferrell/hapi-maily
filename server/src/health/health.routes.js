import boom from 'boom';
import { format } from '../utility';
import { handleCheckServer } from '.';

const createHealthRoutes = () => {
    const notImplemented = (request, reply) => reply(boom.notImplemented());

    const checkServer = {
        method: 'GET',
        path: '/healthCheck',
        handler: handleCheckServer,
        config: {
            auth: false,
            tags: [
                'api', 'Server Utilities'
            ],
            description: 'Get up status of server',
            notes: [
                'Returns success status if the server is online'
            ],
            plugins: {
                'hapi-swagger': {
                    responses: {
                        ...format.healthCheck
                    }
                }
            }
        }
    };

    const checkDependencies = {
        method: 'GET',
        path: '/healthCheck/all',
        handler: notImplemented,
        config: {
            auth: false,
            tags: ['api', 'Server Utilities'],
            description: 'Get up status of server dependencies',
            notes: [
                'Returns an array of each server dependency with up status and latency'
            ],
            plugins: {
                'hapi-swagger': {
                    responses: {
                        ...format.checkDependency,
                        ...format.internalError,
                        ...format.notImplemented
                    }
                }
            }
        }
    };

    return [
        checkServer,
        checkDependencies
    ];
};

export default createHealthRoutes;
