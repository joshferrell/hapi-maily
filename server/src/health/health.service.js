
export const makeGetTime = getDate => (startTime = 0) => getDate().valueOf() - startTime;

export const makeCheckStatus = (promiseFunction, name, getDate, log) => async () => {
    const getTime = makeGetTime(getDate);
    const startTime = getTime();

    try {
        await promiseFunction();
        const duration = getTime(startTime);
        return {
            name,
            up: true,
            duration,
            msg: `${name} is up. time to ping ${duration}ms`
        };
    } catch (err) {
        log.error({ err }, `unable to connect to ${name}`);
        const duration = getTime(startTime);

        return {
            name,
            up: false,
            duration,
            msg: `${name} is down. time to error ${duration}ms`
        };
    }
};
