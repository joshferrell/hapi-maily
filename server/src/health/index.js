import createHealthRoutes from './health.routes';

export { makeCheckSmtpStatus } from './health.service';
export * from './health.handler';
export default createHealthRoutes;
