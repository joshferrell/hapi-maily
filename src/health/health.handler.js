import Boom from 'boom';
import { checkSMTPStatus } from './health.service';

export const healthStatus = (request, reply) => reply({
    status: 'OK'
});

export const createDependencyHandler = (checkMailServer = checkSMTPStatus) =>
    (request, reply) => {
        checkMailServer();
        return reply(Boom.notImplemented());
    };
