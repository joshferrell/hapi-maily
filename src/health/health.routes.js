import Joi from 'joi';
import { internalErrorFormat } from '../utilities/swaggerResponse';

import { healthStatus, createDependencyHandler } from './health.handler';

const healthDependencyStatus = createDependencyHandler();

const healthRoutes = [{
    method: 'GET',
    path: '/health',
    handler: healthStatus,
    config: {
        auth: false,
        tags: ['api', 'Server Utilities'],
        description: 'Get Up Status of Server',
        notes: [
            'Returns success status if the server is online'
        ],
        plugins: {
            'hapi-swagger': {
                responses: {
                    200: {
                        description: 'Server Online',
                        schema: Joi
                            .object({
                                status: Joi
                                    .string()
                                    .equal('OK')
                                    .required()
                            })
                            .required()
                    }
                }
            }
        }
    }
},
{
    method: 'GET',
    path: '/health/all',
    handler: healthDependencyStatus,
    config: {
        auth: false,
        tags: ['api', 'Server Utilities'],
        description: 'Get Up Status of Server Dependencies',
        notes: [
            'Returns an array of each server dependency with up status and latency'
        ],
        plugins: {
            'hapi-swagger': {
                responses: {
                    200: {
                        descripton: 'Server Online',
                        schema: Joi
                            .array()
                            .length(1)
                            .unique()
                            .items(Joi.object({
                                name: Joi
                                    .string()
                                    .valid('smtp')
                                    .required(),
                                up: Joi
                                    .boolean()
                                    .required(),
                                duration: Joi
                                    .number()
                                    .required(),
                                msg: Joi
                                    .string()
                                    .required()
                                    .label('Human readable response about dependency status')
                            }))
                            .required()
                    },
                    ...internalErrorFormat
                }
            }
        }
    }
}];

export default healthRoutes;
