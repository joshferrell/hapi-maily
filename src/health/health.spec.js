import healthRoutes from './health.routes';
import { healthStatus, createDependencyHandler } from './health.handler';

describe('healthcheck methods', () => {
    describe('healthcheck routes', () => {
        it('should return an array of routes for health check', () => {
            expect(healthRoutes).toMatchSnapshot();
        });
    });

    describe('healthcheck handler', () => {
        let reply;
        let request;

        beforeEach(() => {
            reply = jest.fn();
            request = jest.fn();
        });

        describe('health status handler', () => {
            it('should reply that the status is okay', () => {
                healthStatus(request, reply);

                expect(reply).toHaveBeenCalled();
                expect(reply).toHaveBeenCalledTimes(1);
                expect(reply).toHaveBeenCalledWith({
                    status: 'OK'
                });
            });
        });

        describe('health dependency status handler', () => {
            it('should reply what the status of the depencency is', async () => {
                const getEmailServerStatus = jest.fn().mockImplementationOnce(
                    () => Promise.resolve(true)
                );

                const healthDependencyStatus = createDependencyHandler(getEmailServerStatus);
                await healthDependencyStatus(request, reply);

                expect(getEmailServerStatus).toHaveBeenCalled();
                expect(getEmailServerStatus).toHaveBeenCalledTimes(1);
                expect(reply).toHaveBeenCalled();
                expect(reply).toHaveBeenCalledTimes(1);
            });
        });
    });
});
