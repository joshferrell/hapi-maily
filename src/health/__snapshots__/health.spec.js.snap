// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`healthcheck methods healthcheck handler health dependency status handler should throw 500 if a fatal error occurs 1`] = `
Array [
  [Error: Internal Server Error],
]
`;

exports[`healthcheck methods healthcheck routes should return an array of routes for health check 1`] = `
Array [
  Object {
    "config": Object {
      "auth": false,
      "description": "Get Up Status of Server",
      "notes": Array [
        "Returns success status if the server is online",
      ],
      "plugins": Object {
        "hapi-swagger": Object {
          "responses": Object {
            "200": Object {
              "description": "Server Online",
              "schema":  {
                "_baseType": undefined,
                "_description": null,
                "_examples": Array [],
                "_flags": Object {
                  "presence": "required",
                },
                "_inner": Object {
                  "children": Array [
                    Object {
                      "key": "status",
                      "schema":  {
                        "_baseType": undefined,
                        "_description": null,
                        "_examples": Array [],
                        "_flags": Object {
                          "allowOnly": true,
                          "presence": "required",
                        },
                        "_inner": Object {},
                        "_invalids": Set {
                          "_set": Array [
                            "",
                          ],
                        },
                        "_meta": Array [],
                        "_notes": Array [],
                        "_refs": Array [],
                        "_settings": null,
                        "_tags": Array [],
                        "_tests": Array [],
                        "_type": "string",
                        "_unit": null,
                        "_valids": Set {
                          "_set": Array [
                            "OK",
                          ],
                        },
                        "isJoi": true,
                      },
                    },
                  ],
                  "dependencies": Array [],
                  "patterns": Array [],
                  "renames": Array [],
                },
                "_invalids": Set {
                  "_set": Array [],
                },
                "_meta": Array [],
                "_notes": Array [],
                "_refs": Array [],
                "_settings": null,
                "_tags": Array [],
                "_tests": Array [],
                "_type": "object",
                "_unit": null,
                "_valids": Set {
                  "_set": Array [],
                },
                "isJoi": true,
              },
            },
          },
        },
      },
      "tags": Array [
        "api",
        "Server Utilities",
      ],
    },
    "handler": [Function],
    "method": "GET",
    "path": "/health",
  },
  Object {
    "config": Object {
      "auth": false,
      "description": "Get Up Status of Server Dependencies",
      "notes": Array [
        "Returns an array of each server dependency with up status and latency",
      ],
      "plugins": Object {
        "hapi-swagger": Object {
          "responses": Object {
            "200": Object {
              "descripton": "Server Online",
              "schema":  {
                "_baseType": undefined,
                "_description": null,
                "_examples": Array [],
                "_flags": Object {
                  "presence": "required",
                  "sparse": false,
                },
                "_inner": Object {
                  "exclusions": Array [],
                  "inclusions": Array [
                     {
                      "_baseType": undefined,
                      "_description": null,
                      "_examples": Array [],
                      "_flags": Object {},
                      "_inner": Object {
                        "children": Array [
                          Object {
                            "key": "name",
                            "schema":  {
                              "_baseType": undefined,
                              "_description": null,
                              "_examples": Array [],
                              "_flags": Object {
                                "allowOnly": true,
                                "presence": "required",
                              },
                              "_inner": Object {},
                              "_invalids": Set {
                                "_set": Array [
                                  "",
                                ],
                              },
                              "_meta": Array [],
                              "_notes": Array [],
                              "_refs": Array [],
                              "_settings": null,
                              "_tags": Array [],
                              "_tests": Array [],
                              "_type": "string",
                              "_unit": null,
                              "_valids": Set {
                                "_set": Array [
                                  "smtp",
                                ],
                              },
                              "isJoi": true,
                            },
                          },
                          Object {
                            "key": "up",
                            "schema":  {
                              "_baseType": undefined,
                              "_description": null,
                              "_examples": Array [],
                              "_flags": Object {
                                "insensitive": true,
                                "presence": "required",
                              },
                              "_inner": Object {
                                "falsySet": Set {
                                  "_set": Array [],
                                },
                                "truthySet": Set {
                                  "_set": Array [],
                                },
                              },
                              "_invalids": Set {
                                "_set": Array [],
                              },
                              "_meta": Array [],
                              "_notes": Array [],
                              "_refs": Array [],
                              "_settings": null,
                              "_tags": Array [],
                              "_tests": Array [],
                              "_type": "boolean",
                              "_unit": null,
                              "_valids": Set {
                                "_set": Array [],
                              },
                              "isJoi": true,
                            },
                          },
                          Object {
                            "key": "duration",
                            "schema":  {
                              "_baseType": undefined,
                              "_description": null,
                              "_examples": Array [],
                              "_flags": Object {
                                "presence": "required",
                              },
                              "_inner": Object {},
                              "_invalids": Set {
                                "_set": Array [
                                  Infinity,
                                  -Infinity,
                                ],
                              },
                              "_meta": Array [],
                              "_notes": Array [],
                              "_refs": Array [],
                              "_settings": null,
                              "_tags": Array [],
                              "_tests": Array [],
                              "_type": "number",
                              "_unit": null,
                              "_valids": Set {
                                "_set": Array [],
                              },
                              "isJoi": true,
                            },
                          },
                          Object {
                            "key": "msg",
                            "schema":  {
                              "_baseType": undefined,
                              "_description": null,
                              "_examples": Array [],
                              "_flags": Object {
                                "label": "Human readable response about dependency status",
                                "presence": "required",
                              },
                              "_inner": Object {},
                              "_invalids": Set {
                                "_set": Array [
                                  "",
                                ],
                              },
                              "_meta": Array [],
                              "_notes": Array [],
                              "_refs": Array [],
                              "_settings": null,
                              "_tags": Array [],
                              "_tests": Array [],
                              "_type": "string",
                              "_unit": null,
                              "_valids": Set {
                                "_set": Array [],
                              },
                              "isJoi": true,
                            },
                          },
                        ],
                        "dependencies": Array [],
                        "patterns": Array [],
                        "renames": Array [],
                      },
                      "_invalids": Set {
                        "_set": Array [],
                      },
                      "_meta": Array [],
                      "_notes": Array [],
                      "_refs": Array [],
                      "_settings": null,
                      "_tags": Array [],
                      "_tests": Array [],
                      "_type": "object",
                      "_unit": null,
                      "_valids": Set {
                        "_set": Array [],
                      },
                      "isJoi": true,
                    },
                  ],
                  "items": Array [
                     {
                      "_baseType": undefined,
                      "_description": null,
                      "_examples": Array [],
                      "_flags": Object {},
                      "_inner": Object {
                        "children": Array [
                          Object {
                            "key": "name",
                            "schema":  {
                              "_baseType": undefined,
                              "_description": null,
                              "_examples": Array [],
                              "_flags": Object {
                                "allowOnly": true,
                                "presence": "required",
                              },
                              "_inner": Object {},
                              "_invalids": Set {
                                "_set": Array [
                                  "",
                                ],
                              },
                              "_meta": Array [],
                              "_notes": Array [],
                              "_refs": Array [],
                              "_settings": null,
                              "_tags": Array [],
                              "_tests": Array [],
                              "_type": "string",
                              "_unit": null,
                              "_valids": Set {
                                "_set": Array [
                                  "smtp",
                                ],
                              },
                              "isJoi": true,
                            },
                          },
                          Object {
                            "key": "up",
                            "schema":  {
                              "_baseType": undefined,
                              "_description": null,
                              "_examples": Array [],
                              "_flags": Object {
                                "insensitive": true,
                                "presence": "required",
                              },
                              "_inner": Object {
                                "falsySet": Set {
                                  "_set": Array [],
                                },
                                "truthySet": Set {
                                  "_set": Array [],
                                },
                              },
                              "_invalids": Set {
                                "_set": Array [],
                              },
                              "_meta": Array [],
                              "_notes": Array [],
                              "_refs": Array [],
                              "_settings": null,
                              "_tags": Array [],
                              "_tests": Array [],
                              "_type": "boolean",
                              "_unit": null,
                              "_valids": Set {
                                "_set": Array [],
                              },
                              "isJoi": true,
                            },
                          },
                          Object {
                            "key": "duration",
                            "schema":  {
                              "_baseType": undefined,
                              "_description": null,
                              "_examples": Array [],
                              "_flags": Object {
                                "presence": "required",
                              },
                              "_inner": Object {},
                              "_invalids": Set {
                                "_set": Array [
                                  Infinity,
                                  -Infinity,
                                ],
                              },
                              "_meta": Array [],
                              "_notes": Array [],
                              "_refs": Array [],
                              "_settings": null,
                              "_tags": Array [],
                              "_tests": Array [],
                              "_type": "number",
                              "_unit": null,
                              "_valids": Set {
                                "_set": Array [],
                              },
                              "isJoi": true,
                            },
                          },
                          Object {
                            "key": "msg",
                            "schema":  {
                              "_baseType": undefined,
                              "_description": null,
                              "_examples": Array [],
                              "_flags": Object {
                                "label": "Human readable response about dependency status",
                                "presence": "required",
                              },
                              "_inner": Object {},
                              "_invalids": Set {
                                "_set": Array [
                                  "",
                                ],
                              },
                              "_meta": Array [],
                              "_notes": Array [],
                              "_refs": Array [],
                              "_settings": null,
                              "_tags": Array [],
                              "_tests": Array [],
                              "_type": "string",
                              "_unit": null,
                              "_valids": Set {
                                "_set": Array [],
                              },
                              "isJoi": true,
                            },
                          },
                        ],
                        "dependencies": Array [],
                        "patterns": Array [],
                        "renames": Array [],
                      },
                      "_invalids": Set {
                        "_set": Array [],
                      },
                      "_meta": Array [],
                      "_notes": Array [],
                      "_refs": Array [],
                      "_settings": null,
                      "_tags": Array [],
                      "_tests": Array [],
                      "_type": "object",
                      "_unit": null,
                      "_valids": Set {
                        "_set": Array [],
                      },
                      "isJoi": true,
                    },
                  ],
                  "ordereds": Array [],
                  "requireds": Array [],
                },
                "_invalids": Set {
                  "_set": Array [],
                },
                "_meta": Array [],
                "_notes": Array [],
                "_refs": Array [],
                "_settings": null,
                "_tags": Array [],
                "_tests": Array [
                  Object {
                    "arg": 1,
                    "func": [Function],
                    "name": "length",
                    "options": undefined,
                  },
                  Object {
                    "arg": Object {},
                    "func": [Function],
                    "name": "unique",
                    "options": undefined,
                  },
                ],
                "_type": "array",
                "_unit": null,
                "_valids": Set {
                  "_set": Array [],
                },
                "isJoi": true,
              },
            },
            "500": Object {
              "description": "Internal Server Error",
              "schema":  {
                "_baseType": undefined,
                "_description": null,
                "_examples": Array [],
                "_flags": Object {
                  "presence": "required",
                },
                "_inner": Object {
                  "children": Array [
                    Object {
                      "key": "statusCode",
                      "schema":  {
                        "_baseType": undefined,
                        "_description": null,
                        "_examples": Array [],
                        "_flags": Object {
                          "allowOnly": true,
                          "presence": "required",
                        },
                        "_inner": Object {},
                        "_invalids": Set {
                          "_set": Array [
                            "",
                          ],
                        },
                        "_meta": Array [],
                        "_notes": Array [],
                        "_refs": Array [],
                        "_settings": null,
                        "_tags": Array [],
                        "_tests": Array [],
                        "_type": "string",
                        "_unit": null,
                        "_valids": Set {
                          "_set": Array [
                            "500",
                          ],
                        },
                        "isJoi": true,
                      },
                    },
                    Object {
                      "key": "error",
                      "schema":  {
                        "_baseType": undefined,
                        "_description": null,
                        "_examples": Array [],
                        "_flags": Object {
                          "allowOnly": true,
                          "presence": "required",
                        },
                        "_inner": Object {},
                        "_invalids": Set {
                          "_set": Array [
                            "",
                          ],
                        },
                        "_meta": Array [],
                        "_notes": Array [],
                        "_refs": Array [],
                        "_settings": null,
                        "_tags": Array [],
                        "_tests": Array [],
                        "_type": "string",
                        "_unit": null,
                        "_valids": Set {
                          "_set": Array [
                            "Internal Server Error",
                          ],
                        },
                        "isJoi": true,
                      },
                    },
                    Object {
                      "key": "message",
                      "schema":  {
                        "_baseType": undefined,
                        "_description": null,
                        "_examples": Array [],
                        "_flags": Object {
                          "allowOnly": true,
                          "presence": "required",
                        },
                        "_inner": Object {},
                        "_invalids": Set {
                          "_set": Array [
                            "",
                          ],
                        },
                        "_meta": Array [],
                        "_notes": Array [],
                        "_refs": Array [],
                        "_settings": null,
                        "_tags": Array [],
                        "_tests": Array [],
                        "_type": "string",
                        "_unit": null,
                        "_valids": Set {
                          "_set": Array [
                            "An internal server error occurred",
                          ],
                        },
                        "isJoi": true,
                      },
                    },
                  ],
                  "dependencies": Array [],
                  "patterns": Array [],
                  "renames": Array [],
                },
                "_invalids": Set {
                  "_set": Array [],
                },
                "_meta": Array [],
                "_notes": Array [],
                "_refs": Array [],
                "_settings": null,
                "_tags": Array [],
                "_tests": Array [],
                "_type": "object",
                "_unit": null,
                "_valids": Set {
                  "_set": Array [],
                },
                "isJoi": true,
              },
            },
          },
        },
      },
      "tags": Array [
        "api",
        "Server Utilities",
      ],
    },
    "handler": [Function],
    "method": "GET",
    "path": "/health/all",
  },
]
`;
